feat: Implement asset path mapping for Cinema 4D scene files

Ahuge opened the pull request:
Fixes: #125 

### What was the problem/requirement? (What/Why)
When using a Linux worker to render a scene created on Windows, assets used in the scene would not get path mapped
### What was the solution? (How)
Implement path mapping on the asset texture files
### What is the impact of this change?
Path mapping now happens on the C4D scene and Windows paths are converted to Linux ones
### How was this change tested?
Testing In Progress on a client farm. -- IN PROGRESS

### Did you run the "Job Bundle Output Tests"? If not, why not? If so, paste the test results here.

### Was this change documented?
No as it implements functionality that was expected

### Is this a breaking change?
Yes I believe, because we are changing functionality?

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

Ahuge committed:
Add logic to pathmap all assets in scene

Using `paramId` instead of hardcoding `c4d.BITMAPSHADER_FILENAME` makes it work with more asset types that may be linked to other attributes.

This `paramId` references what parameter the asset is assigned to within it's owner

Signed-off-by: Alex Hughes <ahughesalex@gmail.com>

joel-wong-aws commented on src/deadline/cinema4d_adaptor/Cinema4DClient/cinema4d_handler.py:
```diff
@@ -50,6 +50,13 @@ def __init__(self, map_path: Callable[[str], str]) -> None:
+        asset_list = []
+        c4d.documents.GetAllAssetsNew(self.doc, allowDialogs=False, lastPath="", assetList=asset_list)
+        for asset in asset_list:
+            asset_owner = asset.get("owner")
```
I gave this code a try but got `None` as the return value for the `.c4d` file's asset owner, leading to an error trying to index into the `None` object on the line below. 

Have you also encountered this error? This could be as simple as adding an `if asset_owner` condition.

If I only apply this owner to `non-None` asset owners, I still appear to get a segmentation fault from https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/issues/118, but it seems like that is out of scope of his change?
```
2024/12/12 15:52:50-06:00 ADAPTOR_OUTPUT: STDERR: /usr/local/lib/python3.9/site-packages/deadline/cinema4d_adaptor/Cinema4DAdaptor/adaptor.sh: line 55: 185665 Segmentation fault      (core dumped) $C4DEXE ${ARGS[*]}
2024/12/12 15:52:50-06:00 openjd_fail: Error encountered while running adaptor: Cinema4D exited early and did not render successfully, please check render logs. Exit code 139
2024/12/12 15:52:50-06:00 ERROR: Entrypoint failed: STDERR: /usr/local/lib/python3.9/site-packages/deadline/cinema4d_adaptor/Cinema4DAdaptor/adaptor.sh: line 55: 185665 Segmentation fault      (core dumped) $C4DEXE ${ARGS[*]}
2024/12/12 15:52:50-06:00 openjd_fail: Error encountered while running adaptor: Cinema4D exited early and did not render successfully, please check render logs. Exit code 139
```

Ahuge replied:
Checking into that, thanks!

Ahuge replied:
Updated with a guard and now using `paramId` instead of the hardcoded attribute

joel-wong-aws commented:
Thanks for the contribution! 

Does this work for any files that require path mapping or only textures?

What cases does this cover that the existing path mapping does not?
```
if self.render_data[c4d.RDATA_PATH]:
    self.render_data[c4d.RDATA_PATH] = self.map_path(self.render_data[c4d.RDATA_PATH])
```

Ahuge commented:
> Thanks for the contribution!
> 
> Does this work for any files that require path mapping or only textures?
> 
> What cases does this cover that the existing path mapping does not?
> 
> ```
> if self.render_data[c4d.RDATA_PATH]:
>     self.render_data[c4d.RDATA_PATH] = self.map_path(self.render_data[c4d.RDATA_PATH])
> ```

So the RDATA_PATH is just the output render path.
This case works for assets (texture files) used in the scene

karthikbekalp commented on src/deadline/cinema4d_adaptor/Cinema4DClient/cinema4d_handler.py:
```diff
@@ -50,6 +50,15 @@ def __init__(self, map_path: Callable[[str], str]) -> None:
+            asset_owner = asset.get("owner")
+            asset_param_id = asset.get("paramId")
+            if asset_owner and asset_param_id:
+                asset_owner[asset.get("paramId")] = self.map_path(asset.get("filename"))
```
Nit: We could re-use `asset_param_id` instead of calling `asset.get("paramId")`?

Ahuge replied:
Good catch, I'll update it

karthikbekalp commented on src/deadline/cinema4d_adaptor/Cinema4DClient/cinema4d_handler.py:
```diff
@@ -50,6 +50,15 @@ def __init__(self, map_path: Callable[[str], str]) -> None:
+            asset_owner = asset.get("owner")
+            asset_param_id = asset.get("paramId")
+            if asset_owner and asset_param_id:
+                asset_owner[asset.get("paramId")] = self.map_path(asset.get("filename"))
```
I'm trying to better understand this code. How does setting the `asset_owner` variable to a path with applied path mapping rules fix the issue? I don't see `asset_owner` referenced elsewhere else in the code.

karthikbekalp replied:
Hmm, nevermind looks like the object from `asset.get("owner")` has its `paramId` modified and that fixes the issue?

Ahuge replied:
So the `asset` in this case is a dictionary containing all the information about the asset as it is used in the scene.
What we care most about are the `owner` (what node/shader/material/etc is using this asset), the `paramId` (which property in the owner that contains this asset), and the `filename` (The filename path on disk).

So `asset.get("paramId")` returns the attribute on the owner we need to modify the path of.
In most cases it seems like it will be `c4d.BITMAPSHADER_FILENAME` (for image files) but their documentation seems to want us to use the paramId, probably because there are cases where our asset is not an image file, or if a different node references the asset and doesn't store it in `c4d.BITMAPSHADER_FILENAME`.

karthikbekalp replied:
Thanks for the detailed explanation.

karthikbekalp reviewed (commented):
Thanks for your contribution! 

Although I've a question about the fix. Feel free to respond whenever you get time. :)

Ahuge committed:
Set our output path mapping

We should be setting the output paths with a mapped path when we first set the data instead of chaging it later in `start_render`

Signed-off-by: Alex Hughes <ahughesalex@gmail.com>

Ahuge committed:
Fallback to start_render data for the frame

Signed-off-by: Alex Hughes <ahughesalex@gmail.com>

karthikbekalp reviewed (changes requested):
The changes look good, Alex. Can you sign off your commit using `git commit --amend -s` so that the GitHub's DCO check passes?

Ahuge commented:
Updated with signoffs.


Ran tests with the [`cube.py`](https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/tree/mainline/job_bundle_output_tests/redshift_textured/scene) scene as well as a client scene.

Both now do output path mapping as well as asset file mapping.
The key to the output path mapping appears to be mapping the output path *before* we set it in the scene.
Cinema4D on Linux appears to look at a Windows path that we are trying to set and realize that is not a valid Linux path and assume it must be relative.

If we path map before we set it in the scene, we end up passing a valid Linux path to Cinema 4D and everything works as expected

karthikbekalp reviewed (changes requested):
Thanks so much Alex. Unfortunately, there are linter errors now. https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/actions/runs/12319762560/job/34387883976?pr=126 

To solve this, you can run `hatch run fmt` and double check with `hatch run lint`. Sorry for the churn.

Ahuge commented:
> Thanks so much Alex. Unfortunately, there are linter errors now. https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/actions/runs/12319762560/job/34387883976?pr=126
> 
> To solve this, you can run `hatch run fmt` and double check with `hatch run lint`. Sorry for the churn.

No problem, thank you for pointing that out.
Done!

karthikbekalp reviewed (dismissed):
Thanks! The changes look good although I see that we have an empty file inserted `adaptor.sh` in this PR. Is this after the formatting and running the linter?

Ahuge commented:
> Thanks! The changes look good although I see that we have an empty file inserted `adaptor.sh` in this PR. Is this after the formatting and running the linter?

That's very peculiar. I change the executable bit on that file but it shouldn't be empty.
And in my codebase it still shows as existing.

Ahuge commented:
Dropping the executable bit commit to make it clearer, sorry

karthikbekalp reviewed (approved):
Thanks for removing the `adaptor.sh` change. Now the changes look good to me. :)

joel-wong-aws commented:
Thanks for all your work on this, Alex! I've tested this (along with https://github.com/aws-deadline/deadline-cloud-for-cinema-4d/pull/127) and it works great :) 

Thank you also for explaining what `c4d.RDATA_PATH` does to me.

I wouldn't consider this a breaking change, since this fixes an existing bug and shouldn't break any existing workflows that aren't already broken, so I've removed the `!` from the title.

Ship it!
